src/core/AutomationPattern.cpp

```diff
@@ -878,33 +1047,67 @@ void AutomationPattern::generateTangents()
 
 
 
-void AutomationPattern::generateTangents( timeMap::const_iterator it,
-							int numToGenerate )
+// We have two tangents, one for the left side of the node and one for the right side
+// of the node (in case we have discrete value jumps in the middle of a curve).
+// If the inValue and outValue of a node are the same, consequently the inTangent and
+// outTangent values of the node will be the same too.
+void AutomationPattern::generateTangents(timeMap::iterator it, int numToGenerate)
 {
+	QMutexLocker m(&m_patternMutex);
+
 	if( m_timeMap.size() < 2 && numToGenerate > 0 )
 	{
-		m_tangents[it.key()] = 0;
+		it.value().setInTangent(0);
+		it.value().setOutTangent(0);
 		return;
 	}
 
 	for( int i = 0; i < numToGenerate; i++ )
 	{
 		if( it == m_timeMap.begin() )
 		{
-			m_tangents[it.key()] =
-					( (it+1).value() - (it).value() ) /
-						( (it+1).key() - (it).key() );
+			// On the first node there's no curve behind it, so we will only calculate the outTangent
+			// and inTangent will be set to 0.
+			float tangent = ((it + 1).value().getInValue() - (it).value().getOutValue()) /
+				((it + 1).key() - (it).key());
+			it.value().setInTangent(0);
+			it.value().setOutTangent(tangent);
 		}
 		else if( it+1 == m_timeMap.end() )
 		{
-			m_tangents[it.key()] = 0;
+			// Previously, the last value's tangent was always set to 0. That logic was kept for both tangents
+			// of the last node
+			it.value().setInTangent(0);
+			it.value().setOutTangent(0);
 			return;
 		}
 		else
 		{
-			m_tangents[it.key()] =
-					( (it+1).value() - (it-1).value() ) /
-						( (it+1).key() - (it-1).key() );
+			// When we are in a node that is in the middle of two other nodes, we need to check if we
+			// have a discrete jump at this node. If we do not, then we can calculate the tangents normally.
+			// If we do have a discrete jump, then we have to calculate the tangents differently for each side
+			// of the curve.
```

Comment:
I don't like this behavior because `(in, out) = (in, in + very_small_value)` doesn't behave like `(in, in)`.

Reply:
True, that might seem odd at first the the slightest movement can change the curve instantly, but I couldn't think of a better behavior. Usually when we have a discrete jump we want to treat both sides of the curve separately and even if we didn't I think the tangent calculation will look a little odd if we use both sides of the node when there are discrete jumps. Any suggestions? Maybe in the future we will allow manual tangent editing, which would be a more appropriate fix for that I guess.

Reply:
I think #4674 is also related in this context.

Reply:
Adding a TODO comment might be sufficient at this moment.